n2v/grid/grider.py

```diff
@@ -0,0 +1,918 @@
+        wfn: psi4.core.{RKS, UKS}
+        """
+
+        try:
+            vpot = wfn.V_potential()
+        except:
+            raise ValueError("Wfn object does not contain a Vpotential object. Try with one from a DFT calculation")
+
+        nblocks = vpot.nblocks()
+        blocks = [vpot.get_block(i) for i in range(nblocks)]
+        npoints = vpot.grid().npoints()
+        points_function = vpot.properties()[0]
+
+        x = np.empty((npoints))
+        y = np.empty((npoints))
+        z = np.empty((npoints))
+        w = np.empty((npoints))
+
+        offset = 0
+        for i_block in blocks:
+            b_points = i_block.npoints()
+            offset += b_points
+
+            x[offset - b_points : offset] = i_block.x().np
+            y[offset - b_points : offset] = i_block.x().np
+            z[offset - b_points : offset] = i_block.x().np
+            w[offset - b_points : offset] = i_block.x().np
+
+        return [x,y,z,w]
+
+
+    #Quantities on Grid
+    def on_grid_ao(self, matrix, grid=None, vpot=None, cubic_grid=False):
+        """
+        Generates a quantity on the grid given its ao representation
+
+        Parameters
+        ----------
+        matrix: np.ndarray
+            Quantity on ao basis
+        grid: np.ndarray
+            Grid where density will be computed
+        cubic_grid: bool    
+            Is the grid meant for a volume plot?
+        vpot: psi4.core.VBase
+            Vpotential object with info about grid
+
+        Returns
+        -------
+        orbitals: np.ndarray
+            Orbitals on the given grid of size (nbasis, npoints, ref)
+
+        """
+
+        if cubic_grid is True and grid is not None:
+            nshape = grid.shape[1]
+            grid, shape = self.generate_mesh(grid)
+        elif cubic_grid is True and grid is None:
+            raise ValueError("'Cubic Grid' requires to explicitly specify grid")
+
+        if vpot is None and grid is not None:
+            blocks, npoints, points_function = self.grid_to_blocks(grid)
+        elif vpot is not None and grid is None:
+            nblocks = vpot.nblocks()
+            blocks = [vpot.get_block(i) for i in range(nblocks)]
+            npoints = vpot.grid().npoints()
+            points_function = vpot.properties()[0]
+
+        psi4_matrix = psi4.core.Matrix.from_array(matrix)
+        if self.ref == 1:
+            points_function.set_pointers(psi4_matrix)
+        elif self.ref == 2:
+            points_function.set_pointers(psi4_matrix, psi4_matrix)
+        matrix_r = np.empty((npoints)) 
+
+        offset = 0
+        for i_block in blocks:
+            points_function.compute_points(i_block)
+            b_points = i_block.npoints()
+            offset += b_points
+            w = np.array(i_block.w())
+            lpos = np.array(i_block.functions_local_to_global())
+            phi = np.array(points_function.basis_values()["PHI"])[:b_points, :lpos.shape[0]]
+
+            l_mat = matrix[(lpos[:, None], lpos)]
```

Comment:
Here you already assumed matrix is of ndim==2. You should put this code in the following if else.